proof/crefine/RISCV64/SchedContext_C.thy

```diff
@@ -1819,6 +1819,400 @@ lemma invokeSchedControl_ConfigureFlags_ccorres:
   done
 
+lemma mode_parseTimeArg_ccorres_foo:
+  "ccorres (\<lambda>a rv. rv = args ! n) ret__unsigned_longlong_'
+     (sysargs_rel args buffer and sysargs_rel_n args buffer n)
+     (\<lbrace>unat \<acute>i = n\<rbrace> \<inter> \<lbrace>\<acute>buffer = option_to_ptr buffer\<rbrace>) []
+     (return ()) (Call mode_parseTimeArg_'proc)"
+  apply (cinit' lift: i_' buffer_')
+   apply (rule ccorres_add_return2)
+   apply (ctac add: getSyscallArg_ccorres_foo'[where n=n and args=args and buffer=buffer])
+     apply (fastforce intro: ccorres_return_C)
+    apply wp
+   apply (vcg exspec=getSyscallArg_modifies)
+  apply clarsimp
+  done
+
+lemma mode_parseTimeArg_ccorres_foo':
+  "ccorres (\<lambda>a rv. rv = ucast (args ! n)) (\<lambda>x. ucast (ret__unsigned_longlong_' x))
+     (sysargs_rel args buffer and sysargs_rel_n args buffer n)
+     (\<lbrace>unat \<acute>i = n\<rbrace> \<inter> \<lbrace>\<acute>buffer = option_to_ptr buffer\<rbrace>) []
+     (return ()) (Call mode_parseTimeArg_'proc)"
+  apply (insert mode_parseTimeArg_ccorres_foo[where args=args and n=n and buffer=buffer])
+  apply (clarsimp simp: ccorres_underlying_def)
+  apply (erule (1) my_BallE)
+  apply clarsimp
+  apply (erule allE, erule allE, erule (1) impE)
+  apply (clarsimp simp: return_def unif_rrel_def split: xstate.splits)
+  done
+
+lemma refillAbsoluteMax_size_helper:
+  "\<lbrakk>valid_cap' cap s; isSchedContextCap cap\<rbrakk>
+   \<Longrightarrow> MIN_REFILLS \<le> refillAbsoluteMax cap \<and> refillAbsoluteMax cap < 2 ^ LENGTH(64)"
+  supply len_bit0[simp del]
+  apply (clarsimp simp: refillAbsoluteMax_def isSchedContextCap_def split: capability.splits)
+  apply (rule conjI)
+   apply (fastforce intro: MIN_REFILLS_refillAbsoluteMax' simp: valid_cap'_def)
+  apply (simp flip: max_num_refills_eq_refillAbsoluteMax'
+               add: max_num_refils_rewrite max_num_refills'_def valid_cap'_def)
+  apply (rename_tac n)
+  apply (clarsimp simp: refillSizeBytes_sizeof maxUntypedSizeBits_def)
+  apply (rule_tac y="2 ^ n" in le_less_trans)
+   apply fastforce
+  apply (simp add: word_less_nat_alt)
+  apply (simp add: len_bit0)
+  done
+
+lemma MIN_BUDGET_def2:
+  "MIN_BUDGET = us_to_ticks (2 * kernelWCET_us)"
+  by (clarsimp simp: MIN_BUDGET_def kernelWCET_ticks_def timer_defs)
+
+lemma decodeSchedControl_ConfigureFlags_ccorres:
+  "interpret_excaps extraCaps' = excaps_map extraCaps \<Longrightarrow>
+   ccorres (intr_and_se_rel \<currency> dc)  (liftxf errstate id (K ()) ret__unsigned_long_')
+     (\<lambda>s. (invs' s \<and> sch_act_wf (ksSchedulerAction s) s
+           \<and> ksCurThread s = thread \<and> ct_active' s
+           \<and> (excaps_in_mem extraCaps (ctes_of s))
+           \<and> (\<forall>v \<in> set extraCaps. s \<turnstile>' fst v)
+           \<and> (\<forall>v \<in> set extraCaps. \<forall>y \<in> zobj_refs' (fst v). ex_nonz_cap_to' y s)
+           \<and> sysargs_rel args buffer s)
+          \<and> isSchedControlCap cp)
```

Comment:
indent looks off (on github at least)

Reply:
This grouping is done intentionally so that I can do a `ccorres_gen_asm` at the beginning of the proof